pkg/encryption/encryption.go

```diff
@@ -242,3 +380,231 @@ func newRandomKey(length int) (string, error) {
+			added = true
+			break
+		}
+	}
+
+	// provider secret box is not yet present in the config so add the whole configuration
+	if !added {
+		secretBoxProvider := configv1.ProviderConfiguration{
+			Secretbox: &configv1.SecretboxConfiguration{
+				Keys: []configv1.Key{
+					{
+						Name:   keyName(1),
+						Secret: newEncryptionKey,
+					},
+				},
+			},
+		}
+		ec.Resources[0].Providers = append([]configv1.ProviderConfiguration{secretBoxProvider}, ec.Resources[0].Providers...)
+	}
+
+	o, err := yaml.Marshal(ec)
+	if err != nil {
+		return err
+	}
+	secret.Data[EncryptionProviderConfig] = o
+	return nil
+}
+
+// removeOldEncryptionKey will either remove legacy aescbc provider config and its key
+// or remove the last encryption key in secretbox provider
+func removeOldEncryptionKey(secret *v1.Secret) error {
+	var ec configv1.EncryptionConfiguration
+	err := yaml.Unmarshal(secret.Data[EncryptionProviderConfig], &ec)
+	if err != nil {
+		return err
+	}
+
+	// try to remove legacy aescbc provider if present
+	for i, p := range ec.Resources[0].Providers {
+		if p.AESCBC != nil {
+			// found AESCBC provider configuration, should be removed
+			ec.Resources[0].Providers = append(ec.Resources[0].Providers[:i], ec.Resources[0].Providers[i+1:]...)
+
+			o, err := yaml.Marshal(ec)
+			if err != nil {
+				return err
+			}
+			secret.Data[EncryptionProviderConfig] = o
+			return nil
+		}
+	}
+
+	// if no aescbc provider present, remove the last key from the secretbox provider
+	for _, p := range ec.Resources[0].Providers {
+		if p.Secretbox != nil {
+			keysCount := len(p.Secretbox.Keys)
+			if keysCount > 1 {
+				// remove the last key from the array only if there are at least 2 keys
+				p.Secretbox.Keys = p.Secretbox.Keys[:keysCount-1]
+			}
+
+			o, err := yaml.Marshal(ec)
+			if err != nil {
+				return err
+			}
+			secret.Data[EncryptionProviderConfig] = o
+		}
+	}
+	return nil
+}
+
+// rewriteAllSecrets will load all secrets from cluster, add an annotation that marks that it has been rewriten
+// and updates them in API
+func rewriteAllSecrets(wcClient ctrlclient.Client, ctx context.Context) error {
+	var allSecrets v1.SecretList
+	err := wcClient.List(ctx, &allSecrets)
+	if err != nil {
+		return err
+	}
+
+	timestamp := time.Now().Format(time.RFC3339)
+
+	for i := range allSecrets.Items {
+		if allSecrets.Items[i].Annotations == nil {
+			allSecrets.Items[i].Annotations = map[string]string{}
+		}
+		allSecrets.Items[i].Annotations[annotation.EncryptionRewriteTimestamp] = timestamp
+
+		err = wcClient.Update(ctx, &allSecrets.Items[i])
+		if apierrors.IsNotFound(err) {
+			// secret was deleted just ignore and fall thru
+		} else if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Service) countMasterNodesWithLatestConfig(ctx context.Context, wcClient ctrlclient.Client, configShake256Sum string) (bool, error) {
```

Comment:
it's odd to have a function called`countSomething` that returns a boolean. I'd either change the name to areAllMasterNodesUsingLatestConfig` or change the return type to the actual number of nodes.

Reply:
right, I think this function was returning int before but I changed the code a little bit. `areAllMasterNodesUsingLatestConfig` it is